core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnTpch.java

```diff
@@ -0,0 +1,487 @@
+                    }
+                    blocks[i] = block.build();
+                }
+            }
+            else {
+                Random r = new Random(1);
+
+                int totalValueCount = Math.max(distinctValuesCount, dictionaryPositions);
+                BlockBuilder allValues = columnType.type.createBlockBuilder(null, totalValueCount);
+                columnType.blockWriter.write(allValues, totalValueCount, channel);
+                Set<Integer> usedValues = nOutOfM(r, distinctValuesCount, totalValueCount);
+
+                Block[] dictionaries = new Block[distinctDictionaries];
+                Set<Integer>[] possibleIndexesSet = new Set[distinctDictionaries];
+                for (int i = 0; i < distinctDictionaries; i++) {
+                    BlockBuilder dictionaryBuilder = columnType.type.createBlockBuilder(null, dictionaryPositions);
+                    possibleIndexesSet[i] = new HashSet<>();
+                    Set<Integer> distinctValues = nOutOfM(r, dictionaryPositions, totalValueCount);
+                    List<Integer> positions = new ArrayList<>(distinctValues);
+                    Collections.shuffle(positions);
+
+                    for (int j = 0; j < dictionaryPositions; j++) {
+                        int position = positions.get(j);
+                        if (usedValues.contains(position)) {
+                            possibleIndexesSet[i].add(j);
+                        }
+                        columnType.type.appendTo(allValues, position, dictionaryBuilder);
+                    }
+                    dictionaries[i] = dictionaryBuilder.build();
+                }
+                List<Integer>[] possibleIndexes = Arrays.stream(possibleIndexesSet).map(x -> x.stream().collect(toList())).toArray(List[]::new);
+
+                for (int i = 0; i < blockCount; i++) {
+                    int[] indexes = new int[positionsPerBlock];
+                    int dictionaryId = r.nextInt(distinctDictionaries);
+                    Block dictionary = dictionaries[dictionaryId];
+                    int possibleValuesCount = possibleIndexes[dictionaryId].size();
+                    for (int j = 0; j < positionsPerBlock; j++) {
+                        indexes[j] = possibleIndexes[dictionaryId].get(r.nextInt(possibleValuesCount));
+                    }
+
+                    blocks[i] = new DictionaryBlock(dictionary, indexes);
+                }
+            }
+            return blocks;
+        }
+
+        private Set<Integer> nOutOfM(Random r, int n, int m)
+        {
+            Set<Integer> usedValues = new HashSet<>();
+            while (usedValues.size() < n) {
+                int left = n - usedValues.size();
+                for (int i = 0; i < left; i++) {
+                    usedValues.add(r.nextInt(m));
+                }
+            }
+            return usedValues;
+        }
+    }
+
+    public static void main(String[] args)
```

Comment:
Can we add the benchmark commit first and include results before/after of this benchmark as well in the "Replace BigArrays with primitive" commit ?